src/goto-instrument/contracts/contracts.cpp

```diff
@@ -1162,38 +1162,143 @@ bool code_contractst::check_frame_conditions_function(const irep_idt &function)
     decorated.get_recursive_function_warnings_count() == 0,
     "Recursive functions found during inlining");
 
-  // restore internal invariants
+  // Clean up possible fake loops that are due to `IF 0!=0 GOTO i` instructions
+  simplify_gotos(function_body, ns);
+
+  // Restore internal coherence in the programs
+  goto_functions.update();
+
+  INVARIANT(
+    is_loop_free(function_body, ns, log),
+    "Loops remain in function '" + id2string(function) +
+      "', assigns clause checking instrumentation cannot be applied.");
+
+  // Create a deep copy of the original goto_function object
+  // and compute static control flow graph information on it
+  goto_functiont goto_function_orig;
+  goto_function_orig.copy_from(goto_function);
+  cfg_infot cfg_info(ns, goto_function_orig);
+  optionalt<cfg_infot> cfg_info_opt{cfg_info};
+
+  // Add formal parameters to write set
+  for(const auto &param : to_code_type(target.type).parameters())
+    assigns.add_to_write_set(ns.lookup(param.get_identifier()).symbol_expr());
+
+  auto instruction_it = function_body.instructions.begin();
+  for(const auto &car : assigns.get_write_set())
+  {
+    auto snapshot_instructions = car.generate_snapshot_instructions();
+    insert_before_swap_and_advance(
+      function_body, instruction_it, snapshot_instructions);
+  };
+
+  // Restore internal coherence in the programs
   goto_functions.update();
 
   // Insert write set inclusion checks.
   check_frame_conditions(
-    function_obj->first,
-    function_obj->second.body,
+    function,
+    function_body,
     instruction_it,
-    function_obj->second.body.instructions.end(),
-    assigns);
+    function_body.instructions.end(),
+    assigns,
+    // skip checks on function parameter assignments
+    true,
+    cfg_info_opt);
 
   return false;
 }
 
+/// Returns true iff the target instruction is tagged with a
+/// 'CONTRACT_PRAGMA_DISABLE_ASSIGNS_CHECK' pragma.
+bool has_disable_assigns_check_pragma(const goto_programt::targett &target)
```

Comment:
Do you really want an iterator that cannot change (`const goto_programt::targett&`) or an iterator so that whatever it points to cannot change (`goto_programt::const_targett`)?

Reply:
Both! Thanks for the tip